---
title: "Differential expression: Chunk vs dissociated"
author: "Ariel Hippen"
date: '2022-07-25'
output: 
  html_document:
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

# DESeq2 pipeline

This project has two aims that focus on comparing the differential expression of bulk RNA-seq. This one assesses the effect of dissociation by comparing paired tumor samples that were sequenced as chunks vs. dissociated and then bulk sequenced. The second will assess the difference between dissociated cells that have been rRNA-depleted vs those that have been 3'/poly-A captured.

For preliminary assessment of the DE results, we will be following the DESeq2 tutorial: https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
(With some references to https://bioconductor.org/packages/devel/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#exploratory-analysis-and-visualization)

```{r packages}
suppressPackageStartupMessages({
  library(data.table)
  library(DESeq2)
  library(vsn)
  library(pheatmap)
  library(RColorBrewer)
  library(PCAtools)
})

source("../config.R")
```

## Load data

```{r}
# Get paths to STAR.counts files for all ribosomal depleted samples
# Note: data_path is a variable loaded from config.R
directory <- paste(data_path, "bulk_tumors", sep = "/")
sampleFiles <- list.files(directory, recursive = TRUE, full.names = TRUE)
sampleFiles <- grep("ReadsPerGene.out.tab", sampleFiles, value = TRUE)
sampleFiles <- grep("ribo", sampleFiles, value = TRUE)
```

DESeq expects a metadata table to pass into the colData part of a SummarizedExperiment object. We'll prep it here.

```{r}
# Pull sample ids out of full path names. Their ids should be one directory in
# the path and the only exclusively numeric one, hence the regex.
sampleNames <- gsub(".*/(\\d+)/.*", "\\1", sampleFiles)
sampleCondition <- ifelse(1:16 %in% grep("chunk", sampleFiles),
                          "chunk", "dissociated")
sampleUnique <- paste(sampleNames, sampleCondition, sep = "_")
colData <- data.frame(id = sampleUnique,
                      sample = sampleNames,
                      fileName = sampleFiles,
                      condition = sampleCondition)
colData$condition <- factor(colData$condition)

colData
```

Now we can load in the files and create a counts matrix. Note that this shouldn't be stranded data, so we'll use column 2 of the STAR counts files.

```{r}
counts <- matrix(nrow = 36601, ncol = nrow(colData))
for (i in 1:nrow(colData)){
  newcounts <- fread(colData$fileName[i])
  newcounts <- newcounts[-c(1:4), ]
  counts[, i] <- newcounts$V2
  if (i == 1) {
	rownames(counts) <- newcounts$V1
  } else{
	  assert(rownames(counts) == newcounts$V1)
  }
}
rm(newcounts); gc()
```

This counts matrix has the genes listed by their Ensembl IDs, which is helpful for uniqueness but bad for readability in the downstream analysis. The easiest way I've found to convert ensembl IDs to gene names for this dataset is by loading in a SingleCellExperiment object from the same experiment, where this mapping is automatically stored.

```{r}
sce <- readRDS("../sce_objects/pooled_clustered_50.rds")
gene_map <- as.data.frame(subset(rowData(sce), select = c("ID", "Symbol")))
gene_map <- gene_map[rownames(counts) == gene_map$ID, ]
rownames(counts) <- gene_map$Symbol
rm(sce); gc()
```

```{r}
# Create DESeq object 
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = colData,
                              #design = ~condition)
                              design = ~ sample + condition)
dds
```

The tutorial recommends doing some basic pre-filtering to speed up computation, estimate the library depth correction factor, and to clean up later visualizations. We'll remove any row that has fewer than 10 reads.

```{r}
keep <- rowSums(counts(dds)) > = 10
dds <- dds[keep, ]
```

## Differential expression

If you don't set the condition factor specifically, it can be hard to tell if A is upregulated compared to B or vice versa. We'll set "chunk_ribo" as the reference and look at how dissociated_ribo is upregulated or downregulated compared to that.

```{r}
dds$condition <- relevel(dds$condition, ref = "chunk")
```

```{r}
# Run differential expression
dds <- DESeq(dds)
res <- results(dds)
res
```

The tutorial says "shrinkage of effect size (LFC estimates) is useful for visualization and ranking of genes. To shrink the LFC, we pass the dds object to the function lfcShrink. We provide the dds object and the name or number of the coefficient we want to shrink."

```{r}
resLFC <- lfcShrink(dds, coef = "condition_dissociated_vs_chunk", type = "apeglm")
resLFC
```

A quick summary of our differential expression results, at both a 0.1 and 0.05 FDR.

```{r}
summary(res)

sum(res$padj  < 0.1, na.rm = TRUE)
```

```{r}
res05 <- results(dds, alpha = 0.05)
summary(res05)

sum(res05$padj < 0.05, na.rm = TRUE)
```
### Plotting results

From tutorial: "In DESeq2, the function plotMA shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. Points will be colored blue if the adjusted p value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down."

```{r}
plotMA(res, ylim = c(-2, 2))
plotMA(resLFC, ylim = c(-2, 2))
```

### Transformations

The DESeq2 authors recommend the rlog method to adjust for heteroskedasticity in experiments with n < 30. We'll check it and the other vst method they recommend for n > 30.

```{r}
vsd <- vst(dds, blind = FALSE)
rld <- rlog(dds, blind = FALSE)
```

The meanSdPlot plots the mean (as ranked values) by standard deviation, if there is heteroskedasticity there should be a flat line across the values, but they say we shouldn't expect it to be perfectly straight.
```{r}
ntd <- normTransform(dds)
meanSdPlot(assay(ntd))
meanSdPlot(assay(vsd))
meanSdPlot(assay(rld))
```
Okay, none of these are *too* heteroskedastic, but the vsd and rlog look slightly better. I'll use this transformation for the sample comparisons.

## Sample comparisons

```{r}
sampleDists <- dist(t(assay(rld)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$condition, rld$sample, sep = "-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

There's not a ton of clear structure here. It's not exclusively clustering by sample or by dissociation status but by some combination of both. Wonder how this will look on a PCA.

```{r}
pcaData <- plotPCA(rld, intgroup = c("condition", "sample"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = sample, shape = condition)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) + 
  coord_fixed()
```

With a couple of exceptions, the samples seem to be largely clustering together, not as much by dissociation status. But it's also worth noting that the purple and blue samples, which do seem to be clustering by dissociation status, had relatively low RIN scores (<7). Maybe a coincidence but maybe not.

For thoroughness, I'm going to check and see if dissociation status is represented in PCs 3 or 4. I'm going to use the PCAtools tutorial (https://bioconductor.org/packages/release/bioc/vignettes/PCAtools/inst/doc/PCAtools.html) for this.

```{r}
p <- pca(assay(rld), metadata = colData(rld), removeVar = 0.1)
screeplot(p, axisLabSize = 18, titleLabSize = 22)
```
```{r}
pairsplot(p, colby = "condition")
```

It looks like PC3 is dissociation status. 

## Conclusions

- There is definitely some differential expression happening between the chunk samples and the dissociated samples, with slightly more differentially upregulated genes in the dissociated cells.
- Our data may benefit from rlog transformation but it doesn't make a huge difference to downstream results.
- At a high level, our samples are clustering more by patient than by dissociation status, but not perfectly, particularly in the lower quality data (the ones with lower RIN scores).
- Dissociation status is definitely informing samples' distance from one another, but it's not the predominant factor (it's roughly the third principal component).

With this information in hand, we'll try to analyze our differential expression results with GSEA.

```{r}
# Save data
saveRDS(dds, file = "chunk_vs_dissociated_data.rds")

# Save results files
saveRDS(res, file = "chunk_vs_dissociated_FDR_0.1.rds")
saveRDS(res05, file = "chunk_vs_dissociated_FDR_0.05.rds")
```
