
# ---------
# Variables
# ---------

SAMPLES = ['2251', '2267', '2283', '2293', '2380', '2428', '2467', '2497']
BULK_TYPES = ['chunk_ribo', 'dissociated_ribo', 'dissociated_polyA']
THREADS = 6

dir_bulk = "../../sc-cancer-hgsc/data/bulk_tumors"
dir_index = "../../sc-cancer-hgsc/data/index"

wildcard_constraints:
	sample="\d+",
	bulk_type="[A-Za-z]+_[A-Za-z]+"

# --------
# Workflow
# --------

rule all:
	'''This is a stub to get DEseq2 running so I can write the R
	scripts with the proper inputs ready.'''
	input:
		expand(dir_bulk + "/{sample}/{bulk_type}/HTSeq/quantified_gene_expression.tsv", sample=SAMPLES, bulk_type=BULK_TYPES)

#rule evaluate_dissociation_effects:
#	'''A script that runs deseq2 on chunk_ribo vs dissociated_ribo'''
#	input:
#	output:
#		something like {sample}_chunk_v_dissociation.txt
#	shell:
#
#rule compare_mRNA_enrichment:
#	'''A script that runs deseq2 on dissociated_ribo vs dissociated_polyA'''
#	input:
#	output:
#		something like {sample}_polyA_v_ribo.txt
#	shell:

rule run_HTseq:
	'''This should be fairly similar to STAR quantification
	but this might make it play nicer with DEseq2. Actually
	go back and confirm that it does look the same'''
	input:
		alignment = dir_bulk + "/{sample}/{bulk_type}/STAR/Aligned.sortedByCoord.out.bam",
		gtf = dir_index + "/refdata-gex-GRCh38-2020-A/genes/genes.gtf"
	output:
		dir_bulk + "/{sample}/{bulk_type}/HTSeq/quantified_gene_expression.tsv"
	shell:
		'''htseq-count \
		-c {output} \
		{input.alignment} \
		{input.gtf}'''

rule sort_bam:
	input:
		dir_bulk + "/{sample}/{bulk_type}/STAR/Aligned.sortedByCoord.out.bam"
	output:
		dir_bulk + "/{sample}/{bulk_type}/STAR/Aligned.sorted.bam"
	shell:
		'''samtools sort {input} -o {output}'''
