---
title: "Differential expression: rRNA depletion vs poly-A capture"
author: "Ariel Hippen"
date: '2022-07-28'
output: 
  html_document:
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false

---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

# DESeq2 pipeline

This project has two aims that focus on comparing the differential expression of bulk RNA-seq. This one assesses the difference between dissociated cells that have been rRNA-depleted vs those that have been 3'/poly-A captured. The previous one assessed the effect of dissociation by comparing paired tumor samples that were sequenced as chunks vs. dissociated and then bulk sequenced.

For preliminary assessment of the DE results, we will be following the DESeq2 tutorial: https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

```{r packages}
suppressPackageStartupMessages({
  library(data.table)
  library(DESeq2)
  library(vsn)
  library(pheatmap)
  library(RColorBrewer)
  library(PCAtools)
  library(testit)
  library(yaml)
})

params <- read_yaml("../../config.yml")
data_path <- params$data_path
local_data_path <- params$local_data_path
samples <- params$samples
```

## Load data

```{r}
# Get paths to STAR.counts files for all dissociated samples
# Note: data_path is loaded from config.R
directory <- paste(data_path, "bulk_tumors", sep = "/")
sampleFiles <- list.files(directory, recursive = TRUE, full.names = TRUE)
sampleFiles <- grep("ReadsPerGene.out.tab", sampleFiles, value = TRUE)
sampleFiles <- grep("dissociated", sampleFiles, value = TRUE)
```

DESeq expects a metadata table to pass into the colData part of a SummarizedExperiment object. We'll prep it here.

```{r}
# Pull sample ids out of full path names. Their ids should be one directory in
# the path and the only exclusively numeric one, hence the regex.
sampleNames <- gsub(".*/(\\d+)/.*", "\\1", sampleFiles)
sampleCondition <- ifelse(1:16 %in% grep("ribo", sampleFiles),
                          "ribo", "polyA")
sampleUnique <- paste(sampleNames, sampleCondition, sep = "_")
colData <- data.frame(id = sampleUnique,
                      sample = sampleNames,
                      fileName = sampleFiles,
                      condition = sampleCondition)
colData$condition <- factor(colData$condition)

colData
```

Now we can load in the files and create a counts matrix. Note that this shouldn't be stranded data, so we'll use column 2 of the STAR counts files.

```{r}
counts <- matrix(nrow = 36601, ncol = nrow(colData))
for (i in 1:nrow(colData)){
  newcounts <- fread(colData$fileName[i])
  newcounts <- newcounts[-c(1:4), ]
  counts[, i] <- newcounts$V2
  if (i == 1){
	  rownames(counts) <- newcounts$V1
  } else {
	  assert(rownames(counts) == newcounts$V1)
  }
}
rm(newcounts); gc()
```

This counts matrix has the genes listed by their Ensembl IDs, which is helpful for uniqueness but bad for readability in the downstream analysis. The easiest way I've found to convert ensembl IDs to gene names for this dataset is by loading in a SingleCellExperiment object from the same experiment, where this mapping is automatically stored.

```{r}
sce_path <- paste(local_data_path, "sce_objects", sep = "/")
sce <- readRDS(paste(sce_path, "pooled_clustered_50.rds", sep = "/"))
gene_map <- as.data.frame(subset(rowData(sce), select = c("ID", "Symbol")))
gene_map <- gene_map[rownames(counts) == gene_map$ID, ]
rownames(counts) <- gene_map$Symbol
rm(sce); gc()
```

```{r}
# Create DESeq object 
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = colData,
                              #design = ~condition)
                              design = ~ sample + condition)
dds
```

The tutorial recommends doing some basic pre-filtering of non- or low-expressed genes to speed up computation, estimate the library depth correction factor, and to clean up later visualizations. 

```{r}
colnames(counts) <- colData$id
melted_counts <- melt(counts)
melted_counts$logcounts <- log10(melted_counts$value+1)

ggplot(melted_counts, aes(x=Var2, y=logcounts)) + geom_boxplot()
```

Okay, in some of the samples at least 25% of the genes are not expressed. Let's go a little more conservative and remove any genes that have fewer than 20 reads total.

```{r}
keep <- rowSums(counts(dds)) >= 20
dds <- dds[keep, ]
```

## Differential expression

If you don't set the condition factor specifically, it can be hard to tell if A is upregulated compared to B or vice versa. We'll set "ribo" as the reference and look at how dissociated_polyA is upregulated or downregulated compared to that.

```{r}
dds$condition <- relevel(dds$condition, ref = "ribo")
```

```{r}
# Run differential expression
dds <- DESeq(dds)
res <- results(dds)
res
```

The tutorial says "shrinkage of effect size (LFC estimates) is useful for visualization and ranking of genes. To shrink the LFC, we pass the dds object to the function lfcShrink. We provide the dds object and the name or number of the coefficient we want to shrink."

```{r}
resLFC <- lfcShrink(dds, coef = "condition_polyA_vs_ribo", type = "apeglm")
resLFC
```

A quick summary of our differential expression results, at both a 0.1 and 0.05 FDR.

```{r}
summary(res)

sum(res$padj  < 0.1, na.rm = TRUE)
```

```{r}
res05 <- results(dds, alpha = 0.05)
summary(res05)

sum(res05$padj < 0.05, na.rm = TRUE)
```

### Plotting results

From tutorial: "In DESeq2, the function plotMA shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. Points will be colored blue if the adjusted p value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down."

```{r}
plotMA(res, ylim = c(-2, 2))
plotMA(resLFC, ylim = c(-2, 2))
```

### Transformations

The DESeq2 authors recommend the rlog method to adjust for heteroskedasticity in experiments with n < 30. We'll check it and the other vst method they recommend for n > 30.

```{r}
vsd <- vst(dds, blind = FALSE)
rld <- rlog(dds, blind = FALSE)
```

The meanSdPlot plots the mean (as ranked values) by standard deviation, if there is heteroskedasticity there should be a flat line across the values, but they say we shouldn't expect it to be perfectly straight.
```{r}
ntd <- normTransform(dds)
meanSdPlot(assay(ntd))
meanSdPlot(assay(vsd))
meanSdPlot(assay(rld))
```
Okay, regular and rlog look okay, vst looks more heteroskedastic now. I'll use rlog for the sample comparisons.

## Sample comparisons

```{r}
sampleDists <- dist(t(assay(rld)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$condition, rld$sample, sep = "-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

The structure looks fairly sample-specific here.

```{r}
x<-cor(counts)
pheatmap(x)
```

Wow! The correlations are highly based on condition, not sample. Let's see how the PCA looks.

```{r}
pcaData <- plotPCA(rld, intgroup = c("condition", "sample"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = sample, shape = condition)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) + 
  coord_fixed()
```

Whoa. PC2 is obviously ribo depletion, it's almost a little disturbing how perfect the shift is. Not sure what PC1 is besides that it's sample-specific. It doesn't seem to have any relationship to RIN scores, though.

For thoroughness, I'm going to check the lower PCs. I'm going to use the PCAtools tutorial (https://bioconductor.org/packages/release/bioc/vignettes/PCAtools/inst/doc/PCAtools.html) for this.

```{r}
p <- pca(assay(rld), metadata = colData(rld), removeVar = 0.1)
screeplot(p, axisLabSize = 18, titleLabSize = 22)
```

```{r, fig.dim=c(12,12)}
pairsplot(p, colby = "condition", pointSize=2)
```

PC1 is still a bit of a mystery, PC2 is mRNA enrichment method, and PC3-5 also seem to be some sort of sample-specific structure.

## Conclusions

- There is definitely differential expression happening between the ribo-depleted and poly-A captured cells, with what seems like more differentially upregulated genes in the poly-A cells.
- Our data may benefit from rlog transformation but it doesn't make a huge difference to downstream results.
- At a high level, our samples are clustering mostly by patient.
- mRNA enrichment method is definitely informing samples' distance from one another, as the second principal component.

With this information in hand, we'll try to analyze our differential expression results with GSEA.

```{r}
# Save data
deseq_path <- paste(local_data_path, "deseq2_output", sep = "/")
saveRDS(dds, file = paste(deseq_path, "ribo_vs_polyA_data.rds", sep = "/"))

# Save results files
saveRDS(res, file = paste(deseq_path, "ribo_vs_polyA_FDR_0.1.rds", sep = "/"))
saveRDS(res05, file = paste(deseq_path, "ribo_vs_polyA_FDR_0.05.rds", sep = "/"))
```
